channel.go

```diff
@@ -0,0 +1,258 @@
+	messagePathFmt = "channels/%s/%s/message"
+	eventPathFmt   = "channels/%s/%s/event"
+)
+
+type Channel struct {
+	id         string
+	_type      string
+	client     *client
+	customData map[string]interface{}
+}
+
+func (ch Channel) formatPath(path string, params ...interface{}) string {
+	if ch.id == "" {
+		return ""
+	}
+
+	params = append([]interface{}{ch._type, ch.id}, params...)
+
+	return fmt.Sprintf(path, params)
+}
+
+func addUserID(hash map[string]interface{}, userID string) map[string]interface{} {
+	hash["user"] = userID
+	return hash
+}
+
+// SendMessage sends a message to this channel
+//
+// message: the Message object
+// userID: the ID of the user that created the message
+func (ch *Channel) SendMessage(message Message, userID string) error {
+	data := map[string]interface{}{
+		"message": addUserID(message.toHash(), userID),
+	}
+
+	return ch.client.makeRequest(http.MethodPost, ch.formatPath(messagePathFmt), nil, data, nil)
+}
+
+// SendEvent sends an event on this channel
+//
+// event: event data, ie {type: 'message.read'}
+// userID: the ID of the user sending the event
+func (ch *Channel) SendEvent(event Event, userID string) error {
+	data := map[string]interface{}{
+		"event": addUserID(event.toHash(), userID),
+	}
+
+	path := ch.formatPath(eventPathFmt)
+	return ch.client.makeRequest(http.MethodPost, path, nil, data, nil)
+}
+
+// SendReaction sends a reaction about a message
+//
+// messageID: the message id
+// reaction: the reaction object, ie {type: 'love'}
+// userID: the ID of the user that created the reaction
+func (ch *Channel) SendReaction(messageID string, reaction Reaction, userID string) error {
+	data := map[string]interface{}{
+		"reaction": addUserID(reaction.toHash(), userID),
+	}
+
+	path := "messages/" + messageID + "/reaction"
+	return ch.client.makeRequest(http.MethodPost, path, nil, data, nil)
+}
+
+// DeleteReaction removes a reaction by user and type
+//
+// messageID: the id of the message from which te remove the reaction
+// reaction_type: the type of reaction that should be removed
+// userID: the id of the user
+func (ch *Channel) DeleteReaction(messageID string, reactionType string, userID string) error {
+
+	path := "messages/" + messageID + "/reaction/" + reactionType
+
+	return ch.client.makeRequest(http.MethodDelete, path, nil, nil, nil)
+}
+
+// Create creates the channel
+//
+// userID: the ID of the user creating this channel
+func (ch *Channel) Create(userID string) (map[string]interface{}, error) {
+	ch.customData["created_by"] = map[string]interface{}{"id": userID}
+
+	options := map[string]interface{}{
+		"watch":    false,
+		"state":    false,
+		"presence": false,
+	}
+
+	return ch.Query(options)
+}
+
+// Query queries the API for this channel, get messages, members or other channel fields
+//
+// options: the query options, check docs on https://getstream.io/chat/docs/
+func (ch *Channel) Query(options map[string]interface{}) (result map[string]interface{}, err error) {
```

Comment:
Looks like a low level call which shouldn't be exported.

Reply:
I'm not sure if it's low-level call

Reply:
The only promise this interface makes is that the keys are strings. Other than that this in–out is about as loose as it gets. How are we supposed to keep things compatible and consistent?